ksqldb-functional-tests/src/test/java/io/confluent/ksql/test/parser/SqlTestReaderTest.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.test.parser;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThrows;
+
+import io.confluent.ksql.parser.ParsingException;
+import io.confluent.ksql.parser.tree.AssertValues;
+import io.confluent.ksql.test.parser.TestDirective.Type;
+import org.junit.Test;
+
+@SuppressWarnings({"unchecked", "rawtypes"})
+public class SqlTestReaderTest {
+
+  @Test
+  public void shouldParseBasicTest() {
+    // Given:
+    final String contents = ""
+        + "--@test: test1\n"
+        + "CREATE STREAM foo (id INT KEY, col1 INT) WITH (kafka_topic='a', value_format='json');\n"
+        + "CREATE STREAM bar AS SELECT * FROM foo;\n"
+        + "INSERT INTO foo (id, col1) VALUES (1, 1);\n"
+        + "ASSERT VALUES bar (id, col1) VALUES (1, 1);\n"
+        ;
+
+    // When:
+    final SqlTestReader reader = SqlTestReader.of(contents);
+
+    // Then:
+    assertThat(reader.hasNext(), is(true));
+    assertThat(reader.next(), is(TestStatement.of(new TestDirective(Type.TEST, "test1"))));
+    assertThat(reader.hasNext(), is(true));
+    assertThat(reader.next().getEngineStatement().getStatementText(), containsString("CREATE STREAM foo"));
+    assertThat(reader.hasNext(), is(true));
+    assertThat(reader.next().getEngineStatement().getStatementText(), containsString("CREATE STREAM bar"));
+    assertThat(reader.hasNext(), is(true));
+    assertThat(reader.next().getEngineStatement().getStatementText(), containsString("INSERT INTO"));
+    assertThat(reader.hasNext(), is(true));
+    assertThat(reader.next().getAssertStatement(), instanceOf(AssertValues.class));
+    assertThat(reader.hasNext(), is(false));
```

Comment:
If `SqlTestReader` was an `Iterable` rather than an `Iterator` then you're tests could be simple:

```suggestion
    assertThat(reader, contains(
       is(TestStatement.of(new TestDirective(Type.TEST, "test1"))),
       is(engineStatement("CREATE STREAM foo (id INT KEY, col1 INT) WITH (kafka_topic='a', value_format='json');")),
       is(engineStatement("CREATE STREAM bar AS SELECT * FROM foo;")),
       ... etc.
    ));
```

Just a thought... not a biggie.

Reply:
not easy to implement it as an `Iterable` given the nature of TokenStreams and Lexers

Reply:
Sure is easy. Just move all that into the iterator!  the Iterable then just storers the data to be parsed, and the iterator does the parsing and iterating.

Though I agree it would be weird to have it as `Iterable`.

BTW, why not just have it parse every line and return it as a `List`, then iterate through the list?   What's the benefit of it parsing as you go?

Reply:
> Sure is easy. Just move all that into the iterator! the Iterable then just storers the data to be parsed, and the iterator does the parsing and iterating.

Fair enough! Though that does feel weird :P 

> BTW, why not just have it parse every line and return it as a List, then iterate through the list? What's the benefit of it parsing as you go?

In my next PR I actually convert it to a list elsewhere. The reason I do it this way is so that I iterate until I find a `--@test` directive and then I start a new list. It just felt more natural to me than building a list then iterating the list to split it into tests. I could also build the list of lists here, but it turned out nicely with an iterator.

I don't really feel too strongly, so I'll keep this like this for now and if it causes headaches i'll make it return a list.